AD generalized plane strain support (#33416)
idaholab/moose: modules/solid_mechanics/doc/content/modules/solid_mechanics/generalized_plane_strain.md

# Generalized Plane Strain

> ## Description

Reviewer: In making sense of the documentation, there are three main pages for GPS:
1) The theory page (this one). I couldn't figure out how anything links to this. Maybe there was a link from the main SM module page before that got removed:
https://mooseframework.inl.gov/modules/solid_mechanics/generalized_plane_strain.html
2) The syntax page for the action:
https://mooseframework.inl.gov/syntax/Physics/SolidMechanics/GeneralizedPlaneStrain/index.html
3) The source page for the action:
https://mooseframework.inl.gov/source/actions/GeneralizedPlaneStrainAction.html

There are also pages for the actual objects that make up the system that are linked to by these pages. This PR adds duplicates of things to these 3 pages. It is potentially desirable to have the material show up in multiple places, but we want to avoid having multiple copies of material to maintain. We need to decide what parts of the material should go where.

It seems to me like maybe we should completely get rid of this page--let's discuss that.

Generalized plane strain extends plane strain by allowing a nonzero constant
strain in the out-of-plane direction [!citep](Adams1967gps). Other formulations
add two rotational degrees of freedom [!citep](Abaqus2014gps) or include the
anticlastic problem associated with out-of-plane shear [!citep](Adams1984gps, Li1999gps).

The [solid mechanics module](solid_mechanics/index.md) implements the form with

> one extra degree of freedom representing the direct out-of-plane strain

Reviewer: ```suggestion
one extra degree of freedom representing the out-of-plane strain
```
I'm not sure what 'direct' means here, and I think it can be deleted.

[!citep](Li2005gps).

## Formulation

For the usual $x$-$y$ model plane, the $z$ axis is the out-of-plane direction.
Under generalized plane strain conditions, the in-plane stresses are not
functions of $z$, and the direct out-of-plane strain is represented by a scalar
variable. The model is solved on a 2D domain.

The formulation also works when the model plane is $x$-$z$ or $y$-$z$. In those
cases the out-of-plane direction is the $y$ axis or $x$ axis, respectively. In
RZ axisymmetry, the out-of-plane generalized plane strain direction is the axial

> $y$ direction.

Reviewer: ```suggestion
With generalized plane strain, the model is solved for a 2D domain. For the usual 
$x$-$y$ model plane, the $z$ axis is the out-of-plane direction. The out-of-plane
strain is represented by a scalar variable.

The formulation also works when the model plane is $x$-$z$ or $y$-$z$. In those
cases the out-of-plane direction is the $y$ axis or $x$ axis, respectively. 

When generalized plane strain is used for axisymmetric models, the solution domain is 1D,
and the out-of-plane strain direction is the axial $y$ direction.
```

\begin{equation}
\epsilon_{zz} = \bar{\epsilon}_{zz}
\end{equation}
Alternatively, a force as the stress resultant $\bar{N}_{z}$ in the $z$-direction can be prescribed. The condition is the equilibrium condition in $z$-direction, given as
\begin{equation}
\label{eqn:xy_equilibrium}
N_{z} = \int_{A}{\sigma_{zz}dA} = \bar{N}_{z}
\end{equation}
The stress resultant $N_{z}$ conjugates with constant strain $\epsilon_{zz}$.

The formulation above shares the in-plane mechanics of a conventional plane
strain problem and adds the scalar degree of freedom for the out-of-plane
direction.

When an out-of-plane pressure $p_o$ is prescribed, the scalar residual is written
as
\begin{equation}
  R_s = \int_A \left( \sigma_{oo} + p_o \right) dA = 0,
\end{equation}
where $o$ is the selected out-of-plane direction. Positive out-of-plane pressure
is applied toward the body. In Cartesian coordinates, $o=x$, $o=y$, and $o=z$
use $\sigma_{xx}$, $\sigma_{yy}$, and $\sigma_{zz}$, respectively. In RZ
axisymmetry, $o=y$, and the scalar residual is integrated with the RZ coordinate
weighting.

### Implementation

> The out-of-plane strain is a scalar variable in the standard system of equations

Reviewer: ```suggestion
The out-of-plane strain is a scalar variable included as an additional unkonwn in the standard system of equations
```

- [Generalized Plane Strain Off-diagonal Kernel](/GeneralizedPlaneStrainOffDiag.md): in-plane displacement variables and scalar out-of-plane strain coupling

- [Strain Calculations](solid_mechanics/Strains.md): in-plane strain calculation and formation of full strain tensor considering the scalar out-of-plane strain

Objects specific to generalized plane strain with automatic differentiation:

- [ADGeneralizedPlaneStrain.md]: scalar out-of-plane equilibrium condition and
  automatic differentiation (AD) coupling terms

The AD variant of this system requires AD stress materials, although the out-of-plane pressure is
provided as a regular `Real` material property, so it is treated as being derivative-free.

> 

Reviewer: ```suggestion
```
This is covered in the pages about the action, and is out-of-place here.

## How to Use Generalized Plane Strain

The [GeneralizedPlaneStrainAction](/GeneralizedPlaneStrainAction.md) can be used to set up a generalized plane strain model. The [QuasiStaticSolidMechanicsPhysics](/QuasiStaticSolidMechanicsPhysics.md) which considers the [GeneralizedPlaneStrainAction](/GeneralizedPlaneStrainAction.md) as Meta-Action can also be used.

The following example input demonstrates usage of the action with AD, and in this
case the action creates the scalar out-of-plane strain variable.

!listing modules/solid_mechanics/test/tests/generalized_plane_strain/generalized_plane_strain_auto_scalar.i block=Physics/SolidMechanics/GeneralizedPlaneStrain

> 

Reviewer: ```suggestion
```
The links above take us to pages with usage specifics. Let's keep that out of this page.

- [Generalized Plane Strain Off-diagonal Kernel](/GeneralizedPlaneStrainOffDiag.md): in-plane displacement variables and scalar out-of-plane strain coupling

- [Strain Calculations](solid_mechanics/Strains.md): in-plane strain calculation and formation of full strain tensor considering the scalar out-of-plane strain

> Objects specific to the AD generalized plane strain path:

Reviewer: ```suggestion
Objects specific to generalized plane strain with automatic differentiation:
```

- [ADGeneralizedPlaneStrain.md]: scalar out-of-plane equilibrium condition and

>   AD coupling terms

Reviewer: ```suggestion
  automatic differentiation (AD) coupling terms
```

> !table id=gps_legacy_objects caption=Legacy generalized plane strain objects created by the action

Reviewer: The non-AD objects shouldn't be considered "legacy"--they have an equal level of support to the AD ones.

| Object | Role |
| --- | --- |
| [GeneralizedPlaneStrainOffDiag.md] | Off-diagonal coupling between in-plane displacement variables and the scalar out-of-plane strain. |
| [GeneralizedPlaneStrain.md] | Scalar out-of-plane equilibrium residual. |

> | [GeneralizedPlaneStrainUserObject.md] | Residual and diagonal Jacobian data for the scalar kernel. |

Reviewer: High-level comment on this PR:

I was expecting to see that the AD and non-AD code has the same structure, but it looks like this just adopts the new structure for AD. I guess that's OK for an intermediate step, but we really would like to have the two be consistent. I'm not sure how much it would break to move the non-AD code to this structure. It should only affect models that directly invoke the GeneralizedPlaneStrainUserObject, which I don't think is generally done. 

Was your thinking that you would have a follow-on PR that makes the non-AD code mimic the structure of the new AD code? I think that's a reasonable plan.

!table id=gps_ad_objects caption=AD generalized plane strain objects created by the action
| Object | Role |
| --- | --- |
| [ADGeneralizedPlaneStrain.md] | Scalar out-of-plane equilibrium residual and AD coupling terms. |

> The AD path requires AD stress materials. [ADGeneralizedPlaneStrain.md] owns the

Reviewer: The AI tools love to talk about "paths", but that comes off unnaturally to me.

scalar residual assembly through `ADKernelScalarBase`, including displaced-mesh
quadrature weights when requested. The out-of-plane pressure material used by
[ADGeneralizedPlaneStrain.md] is a regular `Real` material property, so it is

> treated as derivative-free in the scalar equation.

Reviewer: ```suggestion
The AD variant of this system requires AD stress materials, although the out-of-plane pressure is
provided as a regular `Real` material property, so it is treated as being derivative-free.
```

## How to Use Generalized Plane Strain

The [GeneralizedPlaneStrainAction](/GeneralizedPlaneStrainAction.md) can be used to set up a generalized plane strain model. The [QuasiStaticSolidMechanicsPhysics](/QuasiStaticSolidMechanicsPhysics.md) which considers the [GeneralizedPlaneStrainAction](/GeneralizedPlaneStrainAction.md) as Meta-Action can also be used.

> The following test-spec block shows a modern QuasiStatic small-strain input run

Reviewer: Maybe this whole section doesn't belong in this file, because it's more of a high-level description of the concept rather than a documentation of the action.

in AD mode.

!listing modules/solid_mechanics/test/tests/generalized_plane_strain/tests start=[generalized_plane_strain_small_ad] end=[] include-end=true

The following standalone action input uses AD mode and lets the action create the
missing scalar out-of-plane strain variable.

!listing modules/solid_mechanics/test/tests/generalized_plane_strain/generalized_plane_strain_auto_scalar.i block=Physics/SolidMechanics/GeneralizedPlaneStrain

> 

Reviewer: ```suggestion
The following example input demonstrates usage of the action with AD, and in this case the action
 creates the scalar out-of-plane strain variable.

!listing modules/solid_mechanics/test/tests/generalized_plane_strain/generalized_plane_strain_auto_scalar.i block=Physics/SolidMechanics/GeneralizedPlaneStrain

```
Changes suggested because I don't think we should be pointing to a test spec with CLI arguments in the documentation--just to blocks of actual input files.
